src/monitors/WhammMonitor.v3

```diff
@@ -83,8 +83,83 @@ class WhammMonitor(whamm: Module) extends Monitor {
 			def wf   = wi.functions[bp.head.2.func_index];
-			def args = bindParams(cv, wf, wi, bp.head.1, opcode, false);
-			if (args != null) Instrumentation.insertLocalProbe(module, func.func_index, opcode_pos, WhammProbe.new(wf, args));
+			// check static predicate
+			if (evalPredicate(cv, wi, opcode, bp.head.3)) {
+				def args = bindParams(cv, wf, wi, bp.head.1, opcode, false);
+				if (args != null) Instrumentation.insertLocalProbe(module, func.func_index, opcode_pos, WhammProbe.new(wf, args));
+			}
+		}
+	}
+
+	def evalPredicate(cv: CodeValidator, wi: Instance, op: Opcode, pred: WhammPredicate) -> bool {
+		match (pred) {
+			True => return true;
+			WhammParam(param) => match (param) {
+				Call(target, params) => {
+					def exports = wi.module.exports;
+					def lookupR = FuncDecl.!(lookup(exports, target.image, Strings.equal, null));
+					if (lookupR == null) {
+						err.at(cv.parser.decoder).WhammProbeError("whamm probe", "function in the probe argument not found");
+						return false;
+					}
+
+					def index = lookupR.func_index;
+					def func = wi.functions[index];
+
+					def whammArgs = bindParams(cv, WasmFunction.!(func), wi, params, op, true);
+					if (whammArgs == null) return false;	// propagating error, not a good idea
+					def args = if(whammArgs.length == 0, Values.NONE, Array<Value>.new(whammArgs.length));
+
+					for (i < whammArgs.length) {
+						var v: Value;
+						match (whammArgs[i]) {
+							FrameAccessor => {
+								err.at(cv.parser.decoder).WhammProbeError("whamm probe", "frame accessor used as an argument to a call");
+								return false;
+							}
+							Val(val)      => v = val;
+							Operand(t, i) => { // this error should not happen here (captured in Arg case)
+								err.at(cv.parser.decoder).WhammProbeError("whamm probe", "arg used as an argument to a call");
+								return false;
+							}
+							Local(t, i)   => { // this error should not happen here (captured in Local case)
+								err.at(cv.parser.decoder).WhammProbeError("whamm probe", "local used as an argument to a call");
+								return false;
+							}
+						}
+						args[i] = v;
+					}
+					// typecheck before execute
+					def result = Execute.call(func, args); // XXX: runs on a new stack
+					match (result) {
+						Value(vals)       => {
+							if (vals.length == 0) {
+								err.at(cv.parser.decoder).WhammProbeError("whamm probe", "function in call argument did not return a value (arity 0/static predicate)");
+								return false;
+							}
+							match (vals[0]) {
+								I32(v) => return v != 0;
+								_ => {
+									err.at(cv.parser.decoder).WhammProbeError("whamm probe", "type mismatch on call return type (static predicate)");
+									return false;
+								}
+							}
+						}
+						Throw(throw)      => {
+							err.at(cv.parser.decoder).WhammProbeError("whamm probe", "function in call argument did not return a value (throw/static predicate)");
```

Comment:
I think the error message can be simplified to "static predicate <the func name> threw exception <except>"